crates/crypto/keystore/src/scrypt.rs

```diff
@@ -0,0 +1,274 @@
+use std::thread;
+
+use anyhow::{anyhow, ensure};
+
+use crate::{pbkdf2::pbkdf2, salsa::salsa20_8_core};
+
+fn scrypt_block_mix(block_size: usize, input_output: &mut [u8]) {
+    let mut block_x = [0u8; 64];
+    block_x.copy_from_slice(&input_output[(2 * block_size - 1) * 64..2 * block_size * 64]);
+
+    for block_index in 0..2 * block_size {
+        let block_offset = block_index * 64;
+        for byte_index in 0..64 {
+            block_x[byte_index] ^= input_output[block_offset + byte_index];
+        }
+        salsa20_8_core(&mut block_x);
+        input_output[block_offset..block_offset + 64].copy_from_slice(&block_x);
+    }
+
+    let mut rearranged_blocks = vec![0u8; input_output.len()];
+    for block_index in 0..block_size {
+        rearranged_blocks[block_index * 64..(block_index + 1) * 64]
+            .copy_from_slice(&input_output[block_index * 128..block_index * 128 + 64]);
+        rearranged_blocks[(block_size + block_index) * 64..(block_size + block_index + 1) * 64]
+            .copy_from_slice(&input_output[block_index * 128 + 64..(block_index + 1) * 128]);
+    }
+
+    input_output.copy_from_slice(&rearranged_blocks);
+}
+
+fn scrypt_romix(block_size: usize, input_output: &mut [u8], cost_parameter: usize) {
```

Comment:
```suggestion
fn scrypt_remix(block_size: usize, input_output: &mut [u8], cost_parameter: usize) {
```

did you mean romix? is that supposed to be random_mix or remix?

Reply:
In https://datatracker.ietf.org/doc/html/rfc7914#page-7

It's referred to as "The scryptROMix Algorithm"